### DISCUSSION TITLE
deploy pocketbase to heroku

### DISCUSSION THREAD

**Discussion by bobwatcherx** [pocketbase/pocketbase#222] [Q&A]
I want to deploy
this pocket base to heroku
with docker
but
when it is a docker image and running there is no error,

I can't access localhost:8090

Or is there another way to get my pocketbase online ?

**Comment by ganigeorgiev** [comment]
Hi, could you share your Dockerfile?

You can change the default http address using the `--http` flag. Debug mode also could be enabled with the `--debug` flag. For example:
```sh
./pocketbase serve --http="127.0.0.1:80" --debug
```

**btw. AFAIK, heroku's free plan doesn't offer disk storage, so I'm not sure that there is an easy way to host PocketBase on Heroku.**

**Comment by bobwatcherx** [comment]
docker sucess run , but can not open in browser

**Comment by bobwatcherx** [comment]
my dockerfile
FROM ubuntu
RUN apt update
COPY . /app
WORKDIR /app
EXPOSE 8090
CMD ["./pocketbase","serve","--http=127.0.0.1:80","--debug"]

**Reply by ganigeorgiev** [reply]
The exposed port (`8090`) doesn't match with the port you are starting the application (`80`).

**Reply by ganigeorgiev** [reply]
Additionally, how do you publish the image to heroku - via the cli? Do you have a heroku.yaml?

**Reply by bobwatcherx** [reply]
> Additionally, how do you publish the image to heroku - via the cli? Do you have a heroku.yaml?
EDIT DOCKERFILE
FROM ubuntu
RUN apt update
COPY . /app
WORKDIR /app
CMD ["./pocketbase","serve","--http=127.0.0.1:80","--debug"]
EXPOSE 80


heroku.yml
build:
  docker:
    web: Dockerfile
run:
  web: WHAT COMMND FOR THIS ////

**Reply by ganigeorgiev** [reply]
I'm sorry, I'm not sure that I understand what actually you are trying to do.

You can make sure first that your docker image runs properly:
```
docker build -t your-app .

# change the ports accordingly to your Dockerfile
docker run -p 8090:80 your-app
```

After that, you can follow the heroku docs for a sample go application - https://devcenter.heroku.com/articles/getting-started-with-go#prepare-the-app (the sample app repo could be found at https://github.com/heroku/go-getting-started).

**But even after doing all of this, keep in mind that PocketBase requires access to the local filesystem, which Heroku doesn't offer (as far as I know their free plan uses ephemeral filesystem, so you'll need something like Litestream to sync your database and files to a remote S3 storage).**

**Comment by bobwatcherx** [comment]
only nginx , but not dashboad pocketbase

**Comment by bobwatcherx** [comment]
thanks for your answer
I found
the problem

I've been confused about why 127.0.0.1 doesn't open the pocketbase dashboard in my browser

and the solution

Dockerfile
CMD ["./pocketbase","serve","--http=0.0.0.0:80","--debug"]

**Comment by bobwatcherx** [comment]
@ganigeorgiev  
is there a way to make the port a variable?
i'm having problem in heroku with heroku logs like this :

 listen tcp 0.0.0.0:80: bind: permission denied

**Reply by ganigeorgiev** [reply]
I think this is because you are using port < 1024. Try with different port, eg. `8090` (update your Dockerfile accordingly).

**But once again, keep in mind that Heroku has epehemeral hard drive - https://devcenter.heroku.com/articles/dynos#ephemeral-filesystem and you'll need something like LiteStream to sync your database to an S3 storage.**

**Comment by bobwatcherx** [comment]
![jad](https://user-images.githubusercontent.com/52072811/181178692-91abbb7a-19e2-493b-9086-b20d7ea8d5f7.gif)

not working, I rebuilt the image and re-deploy it :(

do you know where
I can deploy pocketbase online for free
apart from heroku, aws and google cloud
the good one
thanks

"error message heroku logs:
2022-07-27T06:11:53.380351+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch

**Reply by mystique09** [reply]
> "error message heroku logs:
> 2022-07-27T06:11:53.380351+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch.

Hi, try using this "localhost:$PORT". Afaik, heroku provides port for your app to use, so you don't need to add a PORT env variable in config.

**Reply by bobwatcherx** [reply]
> Hi, try using this "localhost:$PORT". Afaik, heroku provides port for your app to use, so you don't need to add a PORT env variable in config.

thanks solved

**Reply by Dharmang92** [reply]
Hi, can you please elaborate what did you do to fix this? I am facing the same issue.

My Dockerfile. Let me know if you need more information.

```
FROM ubuntu

#ARG PB_VERSION=0.7.2
ARG URL=--http=0.0.0.0:$PORT

RUN apt update && apt install -y --no-install-recommends \
    zip \
    unzip

ADD https://github.com/pocketbase/pocketbase/releases/download/v0.7.2/pocketbase_0.7.2_linux_amd64.zip /tmp/pb.zip
RUN unzip /tmp/pb.zip -d /pb/

CMD ["/pb/pocketbase", "serve", "$URL"]
```

Error in heroku logs:
```
2022-09-26T12:34:08.000000+00:00 app[api]: Build succeeded
2022-09-26T12:34:08.109658+00:00 heroku[web.1]: State changed from crashed to starting
2022-09-26T12:34:10.512858+00:00 heroku[web.1]: Starting process with command `/pb/pocketbase serve \$URL`
2022-09-26T12:34:12.090854+00:00 app[web.1]: > Server started at: http://127.0.0.1:8090
2022-09-26T12:34:12.090869+00:00 app[web.1]:   - REST API: http://127.0.0.1:8090/api/
2022-09-26T12:34:12.090875+00:00 app[web.1]:   - Admin UI: http://127.0.0.1:8090/_/
2022-09-26T12:35:11.145916+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch
2022-09-26T12:35:11.210143+00:00 heroku[web.1]: Stopping process with SIGKILL
2022-09-26T12:35:11.397753+00:00 heroku[web.1]: Process exited with status 137
2022-09-26T12:35:11.578408+00:00 heroku[web.1]: State changed from starting to crashed
```

**Reply by jonshipman** [reply]
Heroku requires PORT to be a variable. You're hard coding it to 8090. $PORT will be random on every run.
